File: Siyeon/[01_24] 스택,큐,덱/4949.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {
    bool isFail;
    while(1) {
        stack<char> bracket;
        string input;
        getline(cin, input);
        
        if(input==".") {
            break;
        }
        isFail = false;
        
        for(int i = 0; i<input.size(); i++) {
            char c = input[i];
            if(c == '(' || c == '[') {
                bracket.push(c);
            }
            else if(c == ')') {
                if(bracket.empty() || bracket.top()!='(') {
                    isFail = true;
                    break;
                }
                bracket.pop();
            }
            else if(c == ']') {
                if(bracket.empty() || bracket.top()!='[') {
                    isFail = true;
                    break;
                }
                bracket.pop();
            }
        }
        if(isFail || !bracket.empty()) cout << "no\n";
        else cout << "yes\n";
    }
}

File: Ayeon/[1월 24일] 스택, 큐, 덱/4949.py
#Stack
#1. Stack underflow 
#2. 문자열 탐색이 끝났는데 스택이 비어있지 않음

import sys
while(True):
    n=list(sys.stdin.readline())
    if ((len(n)==2) & (n[0]=='.')):
        break
    s=[]
    f=True
    for i in n:
        if (i=='(') | (i=='['):
            s.append(i)
        elif (i==')') | (i==']'):
            if len(s)==0:
                f=False
                break
            j=s.pop()
            if ((i==')') & (j=='(')) | ((i==']') & (j=='[')):
                continue
            else:
                f=False
                break
    if len(s)!=0:
        f=False
    if f==False:
        print('no')
    else:
        print('yes')